File: app/src/main/java/com/mbg/mbgsupport/demo/java/singleton/DemoSingletonThird.java
package com.mbg.mbgsupport.demo.java.singleton;


/***
 * java方式线程安全懒汉式创建单例的demo
 * created by canwen2003
 */
public class DemoSingletonThird {
    private static final String TAG="DemoSingletonThird";
    private static DemoSingletonThird instance;
    private DemoSingletonThird(){

    }
    public static synchronized DemoSingletonThird getInstance(){
        if (instance==null){
            instance=new DemoSingletonThird();
        }
        return instance;
    }

    public void logInfo(String str){
        System.out.println(TAG+":"+str);
    }
}

File: app/src/main/java/com/mbg/mbgsupport/demo/kotlin/singleton/DemoSingletonThird.kt
package com.mbg.mbgsupport.demo.kotlin.singleton
/***
 * Kotlin方式懒汉式创建单例的demo
 * created by canwen2003
 */

//私有构造器是为了防止外面创建对象实例化
class DemoSingletonThird private constructor(){
    fun logInfo(str: String){
        println("${TAG}:$str")
    }

    companion object{
        private const val TAG="DemoSingletonThird"
        private var instance :DemoSingletonThird?=null
            get() {
                if (field==null){
                    field= DemoSingletonThird();
                }
                return field;
            }

        @JvmStatic
        @Synchronized
        fun get():DemoSingletonThird{
            return instance!!;
        }
    }
}